tests/components/egps/test_switch.py

```diff
@@ -0,0 +1,82 @@
+"""Test the switch functionality."""
+from collections.abc import Generator
+from unittest.mock import MagicMock, patch
+
+import pytest
+
+from homeassistant.components.egps.const import DOMAIN
+from homeassistant.components.switch import (
+    DOMAIN as SWITCH_DOMAIN,
+    SERVICE_TURN_OFF,
+    SERVICE_TURN_ON,
+)
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.const import ATTR_FRIENDLY_NAME, STATE_OFF, STATE_ON, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+
+from tests.common import MockConfigEntry
+
+
+@pytest.fixture(name="mock_get_device")
+def patch_get_device(pyegps_device_mock: MagicMock) -> Generator[MagicMock, None, None]:
+    """Fixture to patch the `get_device` api method."""
+    with patch(
+        "homeassistant.components.egps.get_device", return_value=pyegps_device_mock
+    ) as mock:
+        yield mock
+
+
+async def test_switch_setup_works(
+    hass: HomeAssistant,
+    device_registry: dr.DeviceRegistry,
+    entity_registry: er.EntityRegistry,
+    valid_config_entry: MockConfigEntry,
+    mock_get_device: MagicMock,
+) -> None:
+    """Test a successful setup of device switches."""
+
+    entry = valid_config_entry
+    entry.add_to_hass(hass)
+
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert entry.state == ConfigEntryState.LOADED
+    assert entry.entry_id in hass.data[DOMAIN]
+
+    device_entry = device_registry.async_get_device(
+        identifiers={(DOMAIN, entry.unique_id)}
+    )
+    entries = er.async_entries_for_device(entity_registry, device_entry.id)
+    switches = [entry for entry in entries if entry.domain == Platform.SWITCH]
+    assert len(switches) == 4
+
+    device = mock_get_device.return_value
+    for switch in switches:
+        socket = switch.unique_id.split("_")[-1]
+        assert (
+            hass.states.get(switch.entity_id).attributes[ATTR_FRIENDLY_NAME]
+            == f"{device.name} Socket {socket}"
+        )
```

Comment:
I'm not sure I like this approach, because it depends on the switches being added successfully and feels a bit error prone. Can we maybe switch to parametrized tests and snapshots?

Reply:
Thx, I see the point in making it a parametrized test. I don't get the snapshot suggestion though, isn't a snapshot testing just comparing an object with its provided stored version? I don't see how this helps in the need of setting up the platform and switch entities before testing them, could you explain? Maybe it would be enough to test only one of the switches though.

Reply:
the snapshot test would just be an additon to quickly check the snapshot attributes of the entities (unique id's, state, category, etc.) and if something change compared to when the snapshots were made. You are already checking the unique id of each switch, a snapshot would just also check everything else is what we were expecting

Reply:
ok, I changed the test

Reply:
sorry, maybe I didn't make myself clear enough: What I meant was: Parametrize over the name of the switches. The on and off calls were fine as they were. The parametrization by on and off exceptions doesn't add any value imo, as it is the same code that is called on the switch anywys.

I don't like checking for sensors being there by only checking the length of the list because that seems error prone. Also I'd suggest you take a snapshot of every single switch's `state = hass.states.get(f"switch.{})` and `entry = entity_registry.async_get(state.entity_id)`. If you parametrize by switch name this can be done with minimal code and easily allows us to track changes to the switches.

Reply:
makes more sense like this now, thanks a lot